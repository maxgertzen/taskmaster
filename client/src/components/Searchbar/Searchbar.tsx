import { FC } from 'react';

import { useTaskStore, useViewportStore } from '../../store/store';
import { TaskInput } from '../TaskInput/TaskInput';

interface SearchbarProps {
  selectedListId: string | null;
  onSearchCallback: () => void;
}

export const Searchbar: FC<SearchbarProps> = ({
  selectedListId,
  onSearchCallback,
}) => {
  const searchTerm = useTaskStore((state) => state.searchTerm);
  const setSearchTerm = useTaskStore((state) => state.setSearchTerm);

  const isMobile = useViewportStore((state) => state.isMobile);

  const handleSearch = (text: string) => {
    onSearchCallback();
    setSearchTerm(text);
  };

  const handleReset = () => {
    if (selectedListId) {
      setSearchTerm('');
    }
  };

  return (
    <TaskInput
      isSearch
      withToggle={isMobile}
      value={searchTerm}
      onReset={handleReset}
      onSubmit={handleSearch}
    />
  );
};
