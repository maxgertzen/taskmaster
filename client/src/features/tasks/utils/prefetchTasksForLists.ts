import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
