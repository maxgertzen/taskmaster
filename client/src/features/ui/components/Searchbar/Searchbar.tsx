import { FC } from 'react';

import { useDashboardStore } from '@/shared/store/dashboardStore';
import { useViewportStore } from '@/shared/store/viewportStore';

import { Input } from '..';

interface SearchbarProps {
  placeholder: string;
  selectedListId?: string | null;
  onSearchCallback: () => void;
}

export const Searchbar: FC<SearchbarProps> = ({
  placeholder,
  selectedListId,
  onSearchCallback,
}) => {
  const { searchTerm, setSearchTerm } = useDashboardStore((state) => state);

  const isMobile = useViewportStore((state) => state.isMobile);

  const handleSearch = (text: string) => {
    onSearchCallback();
    setSearchTerm(text);
  };

  const handleReset = () => {
    if (selectedListId) {
      setSearchTerm('');
    }
  };

  return (
    <Input
      placeholder={placeholder}
      isSearch
      withToggle={isMobile}
      value={searchTerm}
      onReset={handleReset}
      onSubmit={handleSearch}
    />
  );
};
