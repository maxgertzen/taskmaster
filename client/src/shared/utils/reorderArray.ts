export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
