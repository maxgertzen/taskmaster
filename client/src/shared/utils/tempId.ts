export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
