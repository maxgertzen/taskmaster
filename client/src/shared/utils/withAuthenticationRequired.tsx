import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
