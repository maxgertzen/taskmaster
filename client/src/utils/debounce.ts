export function debounce<T extends (...args: Parameters<T>) => ReturnType<T>>(
  func: T,
  delay: number
): ((...args: Parameters<T>) => void) & { cancel: () => void } {
  let timeoutId: ReturnType<typeof setTimeout> | null = null;

  const debouncedFunction = (...args: Parameters<T>) => {
    if (timeoutId) clearTimeout(timeoutId);
    timeoutId = setTimeout(() => func(...args), delay);
  };

  debouncedFunction.cancel = () => {
    if (timeoutId) clearTimeout(timeoutId);
  };

  return debouncedFunction;
}
