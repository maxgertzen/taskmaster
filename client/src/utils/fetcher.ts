export type FetcherConfig<K> = {
  url: string;
  method?: 'GET' | 'POST' | 'PUT' | 'DELETE';
  body?: K;
  timeout?: number;
  headers?: HeadersInit;
  urlSearchParams?: string;
};

export const fetcher = async <T, K = undefined>(
  config: FetcherConfig<K>
): Promise<T> => {
  try {
    const {
      method = 'GET',
      body = {},
      headers = {},
      timeout = 5000,
      urlSearchParams,
    } = config;
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const options: RequestInit = {
      method,
      headers: {
        'Content-Type': 'application/json',
        ...headers,
      },
      ...(method !== 'GET' && { body: JSON.stringify(body) }),
      signal: controller.signal,
    };

    if (urlSearchParams) {
      config.url += `?${urlSearchParams}`;
    }

    const response = await fetch(config.url, options);
    clearTimeout(timeoutId);

    if (!response.ok) {
      throw new Error(`Error: ${response.status} ${response.statusText}`);
    }

    return await response.json();
  } catch (error) {
    console.error('Failed to fetch data:', error);
    throw error;
  }
};
