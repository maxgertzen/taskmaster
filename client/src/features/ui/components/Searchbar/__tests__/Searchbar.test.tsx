import { fireEvent } from '@testing-library/dom';
import { act } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterAll } from 'vitest';

import {
  mockDashboardStore,
  mockViewportStore,
  mockIsMobile,
  mockSetSearchTerm,
} from '@/tests/__mocks__/storeMocks';
import { render, screen } from '@/tests/utils';

import { Searchbar } from '../Searchbar';

vi.mock('@/shared/store/dashboardStore', () => ({
  useDashboardStore: mockDashboardStore,
}));

vi.mock('@/shared/store/viewportStore', () => ({
  useViewportStore: mockViewportStore,
}));

describe('Searchbar', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe.each([
    ['desktop', false],
    ['mobile', true],
  ])('on %s view', (_, isMobile) => {
    beforeEach(() => {
      mockIsMobile.mockReturnValue(isMobile);
    });

    it.each([
      ['light', undefined],
      ['dark', 'dark'],
    ])('renders correctly in %s theme', (_, theme) => {
      const { container } = render(
        <Searchbar placeholder='Search' onSearchCallback={mockSetSearchTerm} />,
        { theme: theme as 'light' | 'dark' }
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Search Behavior', () => {
    beforeEach(() => {
      mockIsMobile.mockReturnValue(false);
      vi.useFakeTimers();
    });

    afterAll(() => {
      vi.useRealTimers();
    });

    it('calls onSearchCallback and setSearchTerm when search is triggered', async () => {
      const mockOnSearchCallback = vi.fn();
      render(
        <Searchbar
          placeholder='Search'
          onSearchCallback={mockOnSearchCallback}
        />
      );

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'test search' } });
      });

      await act(async () => {
        vi.runAllTimers();
      });

      expect(mockOnSearchCallback).toHaveBeenCalled();
      expect(mockSetSearchTerm).toHaveBeenCalledWith('test search');
    });

    it('handles onReset when search is submitted with selectedListId', async () => {
      const mockOnSearchCallback = vi.fn();
      render(
        <Searchbar
          placeholder='Search'
          selectedListId='123'
          onSearchCallback={mockOnSearchCallback}
        />
      );

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'test search' } });
        fireEvent.keyDown(input, { key: 'Enter' });
      });

      expect(mockSetSearchTerm).toHaveBeenCalledWith('');
    });

    it('does not reset search term when submitting without selectedListId', async () => {
      const mockOnSearchCallback = vi.fn();
      render(
        <Searchbar
          placeholder='Search'
          onSearchCallback={mockOnSearchCallback}
        />
      );

      const input = screen.getByTestId('input');

      await act(async () => {
        fireEvent.change(input, { target: { value: 'test search' } });
        fireEvent.keyDown(input, { key: 'Enter' });
      });

      expect(mockSetSearchTerm).not.toHaveBeenCalledWith('');
    });
  });
});
