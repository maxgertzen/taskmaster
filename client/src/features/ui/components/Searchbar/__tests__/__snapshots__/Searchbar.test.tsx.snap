// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Searchbar > on desktop view > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1egpccx"
    data-testid="searchbar"
  >
    <input
      class="css-1q2qwsy"
      data-testid="input"
      placeholder="Search"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Searchbar > on desktop view > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1egpccx"
    data-testid="searchbar"
  >
    <input
      class="css-10md7mk"
      data-testid="input"
      placeholder="Search"
      type="text"
      value=""
    />
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Searchbar > on mobile view > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-1egpccx"
    data-testid="searchbar"
  >
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Searchbar > on mobile view > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-1egpccx"
    data-testid="searchbar"
  >
    <div
      class="css-1rqeb1j"
      data-spotlight=""
    >
      <div
        class="css-y8t6if"
        data-testid="search-icon"
        role="button"
        size="4"
      >
        <svg
          data-testid="magnifying-icon"
        />
      </div>
    </div>
  </div>
</div>
`;
